MusicNotationKit/MusicNotationKit/Tuplet.swift

```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
+        let replacingTicks = noteCollections.reduce(0) { prev, currentCollection in
+            return prev + currentCollection.noteDuration.ticks * currentCollection.noteTimingCount
+        }
+        let toReplaceNote = try note(at: index)
+        let toReplaceTicks = toReplaceNote.noteDuration.ticks * toReplaceNote.noteTimingCount
+        return toReplaceTicks == replacingTicks
     }
 
-    public mutating func replaceNotes(in range: Range<Int>, with tuplet: Tuplet) throws {
-        throw NSError()
+    private func isFlatIndex(_ first: [Int], sameTupletAs second: [Int]) -> Bool {
+        // Same tuplet: counts are equal; nextToLast indexes are equal
+        if first.count == second.count && nextToLastIndex(from: first) == nextToLastIndex(from: second) {
+            return true
+        } else {
+            return false
+        }
     }
 
-    // MARK: Private
+    private func nextToLastIndex(from flatIndex: [Int]) -> Int? {
+        // Array.endIndex == count
+        // Is this the best way to get second to last index?
+        return flatIndex.count > 1 ? flatIndex[flatIndex.endIndex - 2] : nil
+    }
 
-    internal mutating func replaceNote(at index: Int, with noteCollection: NoteCollection) throws {
-        // validate they are the same duration
-        let noteToReplace = try note(at: index)
-        let noteCollectionDuration = NoteDuration.number(
-            of: noteToReplace.noteDuration,
-            within: noteCollection.noteDuration) *
-            Double(noteCollection.noteTimingCount)
-        guard noteCollectionDuration == 1 else {
-            throw TupletError.replacingCollectionNotSameDuration
+    private func isValidReplacementRange(_ range: CountableClosedRange<Int>) throws -> Bool {
+        guard range.lowerBound >= 0 && range.lowerBound < flatIndexes.count
+            && range.upperBound >= 0 && range.upperBound < flatIndexes.count else {
+                throw TupletError.invalidIndex
         }
-        let fullIndexes = flatIndexes[index]
-        try replaceNote(at: fullIndexes, with: noteCollection)
+        let firstFlatIndex = flatIndexes[range.lowerBound]
+        let lastFlatIndex = flatIndexes[range.upperBound]
+
+        // first & last are in same tuplet
+        if isFlatIndex(firstFlatIndex, sameTupletAs: lastFlatIndex) {
+            return true
+        }
+
+        func flatIndex(at index: Int) -> [Int]? {
+            guard index >= 0 && index < flatIndexes.count else {
+                return nil
+            }
+            return flatIndexes[index]
+        }
+
+        // If first - 1 (tuplet) != first (tuplet) && last + 1 (tuplet) != last (tuplet): return true;
+        // OR the count is 1 (not a nested tuplet)
+        // else return false
+        let beforeFirstFlatIndex = flatIndex(at: range.lowerBound - 1)
+        let afterLastFlatIndex = flatIndex(at: range.upperBound + 1)
+
+        switch (beforeFirstFlatIndex, afterLastFlatIndex) {
```

Comment:
Hmm. I tried what you are saying and it doesn't seem to cover everything correctly as the tests pass. Do you find this particularly difficult to reason through? Or do you think it's overly complicated? I think the switch statement and comment above makes it not too hard to follow.